### PR TITLE
fix(edge): error handling for edge route and middleware is inconsistent

### PR DESCRIPTION
**Pull request by feugy**
## What’s in there?

This PR brings more consistency in how errors and warnings are reported when running code in the Edge Runtime:

- Dynamic code evaluation (`eval()`, `new Function()`, `WebAssembly.instantiate()`, `WebAssembly.compile()`…)
- Usage of Node.js global APIs (`BroadcastChannel`, `Buffer`, `TextDecoderStream`, `setImmediate()`...)
- Usage of Node.js modules (`fs`, `path`, `child_process`…)

The new error messages should mention *Edge Runtime* instead of *Middleware*, so they are valid in both cases.

It also fixes a bug where the process polyfill would issue a warning for  `process.cwd` (which is `undefined` but legit). Now, one has to invoke the function `process.cwd()` to trigger the error.

It finally fixes the react-dev-overlay, where links from middleware and Edge API route files could not be opened because of the `(middleware)/` prefix in their name.

About the later, please note that we can’t easily remove the prefix or change it for Edge API routes. It comes from the Webpack layer, which is the same for both. We may consider renaming it to *edge* instead in the future.

## How to test?

These changes are almost fully covered with tests:

```bash
pnpm testheadless --testPathPattern runtime-dynamic
pnpm testheadless --testPathPattern runtime-with-node
pnpm testheadless --testPathPattern runtime-module
pnpm testheadless --testPathPattern middleware-dev-errors
```

To try them out manually, you can write a middleware and Edge route files like these:

```jsx
// middleware.js
import { NextResponse } from 'next/server'
import { basename } from 'path'

export default async function middleware() {
  eval('2+2')
  setImmediate(() => {})
  basename()
  return NextResponse.next()
}

export const config = { matcher: '/' }
```

```jsx
// pages/api/route.js
import { basename } from 'path'

export default async function handle() {
  eval('2+2')
  setImmediate(() => {})
  basename()
  return Response.json({ ok: true })
}

export const config = { runtime: 'experimental-edge' }
```

The expected behaviours are:

- [x] dev, middleware/edge route is using a node.js module: error at runtime (logs + read-dev-overlay):

```bash
error - (middleware)/pages/api/route.js (1:0) @ Object.handle [as handler]
error - The edge runtime does not support Node.js 'path' module.
Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime
> 1 | import { basename } from "path";
  2 | export default async function handle() {
```

- [x] build, middleware/edge route is using a node.js module: warning but succeeds

```bash
warn  - Compiled with warnings

./middleware.js
A Node.js module is loaded ('path' at line 4) which is not supported in the Edge Runtime.
Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime

./pages/api/route.js
A Node.js module is loaded ('path' at line 1) which is not supported in the Edge Runtime.
Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime
```

- [x] production, middleware/edge route is using a node.js module: error at runtime (logs + 500 error)

```bash
Error: The edge runtime does not support Node.js 'path' module.
Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime
    at <unknown> (file:///Users/damien/dev/next.js/packages/next/server/web/sandbox/context.ts:149)
```

- [x] dev, middleware/edge route is using a node.js global API: error at runtime (logs + read-dev-overlay):

```bash
error - (middleware)/pages/api/route.js (4:2) @ Object.handle [as handler]
error - A Node.js API is used (setImmediate) which is not supported in the Edge Runtime.
Learn more: https://nextjs.org/docs/api-reference/edge-runtime
  2 |
  3 | export default async function handle() {
> 4 |   setImmediate(() => {})
    |  ^
```

- [x] build, middleware/edge route is using a node.js global API: warning but succeeds

```bash
warn  - Compiled with warnings

./middleware.js
A Node.js API is used (setImmediate at line: 6) which is not supported in the Edge Runtime.
Learn more: https://nextjs.org/docs/api-reference/edge-runtime

./pages/api/route.js
A Node.js API is used (setImmediate at line: 3) which is not supported in the Edge Runtime.
Learn more: https://nextjs.org/docs/api-reference/edge-runtime
```

- [x] production, middleware/edge route is using a node.js module: error at runtime (logs + 500 error)

```bash
Error: A Node.js API is used (setImmediate) which is not supported in the Edge Runtime.
Learn more: https://nextjs.org/docs/api-reference/edge-runtime
    at <unknown> (file:///Users/damien/dev/next.js/packages/next/server/web/sandbox/context.ts:330)
```

- [x] dev, middleware/edge route is loading dynamic code: warning at runtime (logs + read-dev-overlay) and request succeeds (we allow dynamic code in dev only):

```bash
warn  - (middleware)/middleware.js (7:2) @ Object.middleware [as handler]
warn  - Dynamic Code Evaluation (e. g. 'eval', 'new Function') not allowed in Edge Runtime
   5 |
   6 | export default async function middleware() {
>  7 |   eval('2+2')
```

- [x] build, middleware/edge route is loading dynamic code: build fails with error:

```bash
Failed to compile.

./middleware.js
Dynamic Code Evaluation (e. g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime
Used by default

./pages/api/route.js
Dynamic Code Evaluation (e. g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime
Used by default
```

## Notes to reviewers

Edge-related errors are either issued from `next/server/web/sandbox/context.ts` file (runtime errors) or from `next/build/webpack/plugins/middleware-plugin.ts` (webpack compilation).

The previous implementation (I’m pleading guilty here) was way too verbose: some errors (Node.js global APIs like using `process.cwd()`) could be reported several times, and the previous mechanism to dedupe them (in middleware-plugin) wasn’t really effective.

Changes in tests are due to renaming existing tests such as `test/integration/middleware-with-node.js-apis` into `test/integration/edge-runtime-with-node.js-apis`. I extended them to cover Edge API route.

@hanneslund I’ve pushed the improvement you did in https://github.com/vercel/next.js/pull/38289/ one step further to avoid duplication.

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: e42bdf7d6f5eccb1df9efb3faf3790cad5b6fb18

`pnpm testheadless test/integration/react-server-components/test/switchable-runtime.test.js` 
- Switchable runtime (prod) > should build /edge as a dynamic page with the edge runtime
- Switchable runtime (prod) > should build /edge-rsc as a dynamic page with the edge runtime

<details>
<summary>Expand output</summary>

  ● Switchable runtime (prod) › should build /edge as a dynamic page with the edge runtime

    expect(received).toContain(expected) // indexOf

    Expected substring: "Runtime: Edge"
    Received string:    "<!DOCTYPE html><html><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width\"/><meta name=\"next-head-count\" content=\"2\"/><meta name=\"next-font-preconnect\"/><noscript data-n-css=\"\"></noscript><link rel=\"preload\" href=\"/_next/static/chunks/webpack-69bfa6990bb9e155.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/framework-75ec34940eea6891.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/main-265ce40130021fa9.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/pages/_app-fa072057c4e3e5a5.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/pages/edge-e4bdedeb991f5a5c.js\" as=\"script\"/></head><body><div id=\"__next\"><div class=\"app-client-root\" data-title=\"\"><div>This is a SSR page.<br/><span id=\"__runtime\">Runtime: Node.js ()=&gt;emitError(`process.${key}`)</span><br/><span id=\"__time\">Time: 1657182878482</span></div></div></div><script id=\"__NEXT_DATA__\" type=\"application/json\">{\"props\":{\"pageProps\":{}},\"page\":\"/edge\",\"query\":{},\"buildId\":\"pWrgDf5yXBUCSXHYYv6dn\",\"autoExport\":true,\"isFallback\":false,\"customServer\":true,\"scriptLoader\":[]}</script><script nomodule=\"\" src=\"/_next/static/chunks/polyfills-0d1b80a048d4787e.js\"></script><script src=\"/_next/static/chunks/webpack-69bfa6990bb9e155.js\" async=\"\"></script><script src=\"/_next/static/chunks/framework-75ec34940eea6891.js\" async=\"\"></script><script src=\"/_next/static/chunks/main-265ce40130021fa9.js\" async=\"\"></script><script src=\"/_next/static/chunks/pages/_app-fa072057c4e3e5a5.js\" async=\"\"></script><script src=\"/_next/static/chunks/pages/edge-e4bdedeb991f5a5c.js\" async=\"\"></script><script src=\"/_next/static/pWrgDf5yXBUCSXHYYv6dn/_buildManifest.js\" async=\"\"></script><script src=\"/_next/static/pWrgDf5yXBUCSXHYYv6dn/_ssgManifest.js\" async=\"\"></script></body></html>"

      52 |   const html1 = await renderViaHTTP(appPort, url)
      53 |   const renderedAt1 = +html1.match(/Time: (\d+)/)[1]
    > 54 |   expect(html1).toContain(`Runtime: ${isEdge ? 'Edge' : 'Node.js'}`)
         |                 ^
      55 |
      56 |   const html2 = await renderViaHTTP(appPort, url)
      57 |   const renderedAt2 = +html2.match(/Time: (\d+)/)[1]

      at testRoute (integration/react-server-components/test/switchable-runtime.test.js:54:17)
      at Object.<anonymous> (integration/react-server-components/test/switchable-runtime.test.js:165:5)

  ● Switchable runtime (prod) › should build /edge-rsc as a dynamic page with the edge runtime

    expect(received).toContain(expected) // indexOf

    Expected substring: "Runtime: Edge"
    Received string:    "<!DOCTYPE html><html><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width\"/><meta name=\"next-head-count\" content=\"2\"/><meta name=\"next-font-preconnect\"/><noscript data-n-css=\"\"></noscript><link rel=\"preload\" href=\"/_next/static/chunks/webpack-69bfa6990bb9e155.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/framework-75ec34940eea6891.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/main-265ce40130021fa9.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/pages/_app-fa072057c4e3e5a5.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/pages/edge-rsc-756a72a938552f5e.js\" as=\"script\"/></head><body><div id=\"__next\"><div class=\"app-client-root\" data-title=\"\"><div>This is a SSR RSC page.<br/><span id=\"__runtime\">Runtime: Node.js ()=&gt;emitError(`process.${key}`)</span><br/><span id=\"__time\">Time: 1657182878615</span></div></div></div><script>(self.__next_s=self.__next_s||[]).push([0,\"/edge-rsc,:R6:\"])</script><script>(self.__next_s=self.__next_s||[]).push([1,\"/edge-rsc,:R6:\",\"J0:[\\\"$\\\",\\\"div\\\",null,{\\\"children\\\":[\\\"This is a SSR RSC page.\\\",[\\\"$\\\",\\\"br\\\",null,{}],[\\\"$\\\",\\\"span\\\",null,{\\\"id\\\":\\\"__runtime\\\",\\\"children\\\":\\\"Runtime: Node.js ()=\\u003eemitError(`process.${key}`)\\\"}],[\\\"$\\\",\\\"br\\\",null,{}],[\\\"$\\\",\\\"span\\\",null,{\\\"id\\\":\\\"__time\\\",\\\"children\\\":\\\"Time: 1657182878615\\\"}]]}]\\n\"])</script><script id=\"__NEXT_DATA__\" type=\"application/json\">{\"props\":{\"pageProps\":{}},\"page\":\"/edge-rsc\",\"query\":{},\"buildId\":\"pWrgDf5yXBUCSXHYYv6dn\",\"isFallback\":false,\"rsc\":true,\"customServer\":true,\"scriptLoader\":[]}</script><script nomodule=\"\" src=\"/_next/static/chunks/polyfills-0d1b80a048d4787e.js\"></script><script src=\"/_next/static/chunks/webpack-69bfa6990bb9e155.js\" async=\"\"></script><script src=\"/_next/static/chunks/framework-75ec34940eea6891.js\" async=\"\"></script><script src=\"/_next/static/chunks/main-265ce40130021fa9.js\" async=\"\"></script><script src=\"/_next/static/chunks/pages/_app-fa072057c4e3e5a5.js\" async=\"\"></script><script src=\"/_next/static/chunks/pages/edge-rsc-756a72a938552f5e.js\" async=\"\"></script><script src=\"/_next/static/pWrgDf5yXBUCSXHYYv6dn/_buildManifest.js\" async=\"\"></script><script src=\"/_next/static/pWrgDf5yXBUCSXHYYv6dn/_ssgManifest.js\" async=\"\"></script></body></html>"

      52 |   const html1 = await renderViaHTTP(appPort, url)
      53 |   const renderedAt1 = +html1.match(/Time: (\d+)/)[1]
    > 54 |   expect(html1).toContain(`Runtime: ${isEdge ? 'Edge' : 'Node.js'}`)
         |                 ^
      55 |
      56 |   const html2 = await renderViaHTTP(appPort, url)
      57 |   const renderedAt2 = +html2.match(/Time: (\d+)/)[1]

      at testRoute (integration/react-server-components/test/switchable-runtime.test.js:54:17)
      at Object.<anonymous> (integration/react-server-components/test/switchable-runtime.test.js:172:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by nuta — COMMENTED** [review]
I'm in favor of this improvement. Left some nitpicks.

File: test/integration/middleware-dev-errors/test/index.test.js [hunk 1]
```diff
@@ -124,7 +124,7 @@ describe('Middleware development errors', () => {
       const output = stripAnsi(context.logs.output)
       expect(output).toMatch(
         new RegExp(
-          `error - \\(middleware\\)/middleware.js \\(\\d+:\\d+\\) @ eval\nReferenceError: test is not defined`,
+          `error - \\(middleware\\)/middleware.js \\(\\d+:\\d+\\) @ eval\nerror - test is not defined`,
```
[the remark this answers is not in the document]
Comment (ethomson) [reply test/integration/middleware-dev-errors/test/index.test.js:127]: Personally, your new message feels simpler without losing useful information.

**Review by ijjk — DISMISSED** [review]

**Review by ijjk — COMMENTED** [review]
I'm seeing the below error while testing some changes against front, are we failing to initialize `__import_unsupported` somewhere?

```sh
/api/screenshot
17:11:24:11
ReferenceError: __import_unsupported is not defined
    at external root " __import_unsupported('os')":1:0
    at webpack/bootstrap:21:0
    at node_modules/clean-stack/index.js:2:11
    at webpack/bootstrap:21:0
    at node_modules/whoops/lib/create-extend-error.js:3:19
    at webpack/bootstrap:21:0
    at node_modules/whoops/lib/index.js:3:26
    at webpack/bootstrap:21:0
    at node_modules/@microlink/mql/src/browser.js:6:15
    at webpack/bootstrap:21:0
```

**Comment by feugy** [comment]
@ijjk you're right, and I apologize.
I wrote https://github.com/vercel/next.js/pull/38769 to address the issue.
Although the bug is related to error handling in edge-runtime, it is not related to the current PR.

**Review by ijjk — APPROVED** [review]

**Pull request merged**